motionplan/nloptIKSolver.go

```diff
@@ -22,49 +23,90 @@ var (
-// NloptIK TODO.
-type NloptIK struct {
-	id            int
-	model         referenceframe.Frame
-	lowerBound    []float64
-	upperBound    []float64
-	maxIterations int
-	epsilon       float64
-	solveEpsilon  float64
-	logger        golog.Logger
-	jump          float64
-}
+	// Default distance below which two distances are considered equal.
+	defaultEpsilon = 1e-3
 
-// CreateNloptIKSolver creates an nloptIK object that can perform gradient descent on metrics for Frames. The parameters are the Frame on
-// which Transform() will be called, a logger, and the number of iterations to run. If the iteration count is less than 1, it will be set
-// to the default of 5000.
-func CreateNloptIKSolver(mdl referenceframe.Frame, logger golog.Logger, iter int) (*NloptIK, error) {
-	ik := &NloptIK{logger: logger}
+	defaultSolveEpsilon = 1e-12
+	defaultJump         = 1e-8
+)
+
+type nloptOptions struct {
+	ConstrainedTries  int `json:"constrained_tries"`
+	StepsPerIteration int `json:"steps_per_iteration"`
+	MaxIterations     int `json:"iterations"`
 
-	ik.model = mdl
-	ik.id = 0
 	// How close we want to get to the goal
-	ik.epsilon = defaultEpsilon
+	Epsilon float64 `json:"epsilon"`
+
 	// Stop optimizing when iterations change by less than this much
-	ik.solveEpsilon = math.Pow(ik.epsilon, 4)
-	if iter < 1 {
-		// default value
-		iter = 5000
-	}
-	ik.maxIterations = iter
-	ik.lowerBound, ik.upperBound = limitsToArrays(mdl.DoF())
+	SolveEpsilon float64 `json:"solve_epsilon"`
+
 	// How much to adjust joints to determine slope
-	ik.jump = 0.00000001
+	Jump float64 `json:"jump"`
+}
+
+func newNLOptOptions(opt *ikOptions) (*nloptOptions, error) {
+	algOpts := &nloptOptions{
+		ConstrainedTries:  defaultConstrainedTries,
+		StepsPerIteration: defaultStepsPerIter,
+		MaxIterations:     defaultMaxIterations,
+		Epsilon:           defaultEpsilon,
+		SolveEpsilon:      defaultSolveEpsilon,
+		Jump:              defaultJump,
+	}
+	// convert map to json
+	jsonString, err := json.Marshal(opt.extra)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(jsonString, algOpts)
+	if err != nil {
+		return nil, err
+	}
+	return algOpts, nil
+}
 
+type nloptIKSolver struct {
+	*ikSolver
+	algOpts    *nloptOptions
+	id         int
+	lowerBound []float64
+	upperBound []float64
+}
+
+// newNLOptIKSolver creates an nloptIK object that can perform gradient descent on metrics for Frames. The parameters are the Frame on
+// which Transform() will be called, a logger, and the number of iterations to run. If the iteration count is less than 1, it will be set
+// to the default of 5000.
+func newNLOptIKSolver(model referenceframe.Frame, logger golog.Logger, opts *ikOptions) (*nloptIKSolver, error) {
+	if opts == nil {
+		opts = newBasicIKOptions()
+	}
+	algOpts, err := newNLOptOptions(opts)
+	if err != nil {
+		return nil, err
+	}
+	ik := &nloptIKSolver{
+		ikSolver: &ikSolver{
+			logger: logger,
+			model:  model,
+			opts:   opts,
+		},
+		algOpts: algOpts,
+		id:      0,
```

Comment:
I'm not seeing any place where we change the `id`, despite having logic for id == 1 later on. Is this still necessary?

Reply:
That's inside the ensembleIKsolver

Reply:
Yep, found it, thanks for the pointer.